engine/resources/consul_kv.go

```diff
@@ -0,0 +1,186 @@
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/purpleidea/mgmt/engine"
+	"github.com/purpleidea/mgmt/engine/traits"
+	"github.com/purpleidea/mgmt/util"
+)
+
+func init() {
+	engine.RegisterResource("consul:kv", func() engine.Res { return &ConsulKVRes{} })
+}
+
+// ConsulKVRes is a resource that writes a value into a consul datastore.
+type ConsulKVRes struct {
+	traits.Base
+	init *engine.Init
+
+	// Key is the path of the value.
+	Key string `lang:"key" yaml:"key"`
+
+	// Key is the value that will be written into consul.
+	Value string `lang:"value" yaml:"value"` // the consul key value
+
+	// Address is the address of the consul server.
+	Address string `lang:"address" yaml:"address"` // the consul server
+
+	// Token is the token used to connect to the consul server.
+	Token string `lang:"token" yaml:"token"` // the consul token
+
+	client *api.Client
+	config *api.Config // needed to close the idle connections
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *ConsulKVRes) Default() engine.Res {
+	return &ConsulKVRes{}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *ConsulKVRes) Validate() error {
+	if obj.Key == "" {
+		return errors.New("key is mandatory")
+	}
+	return nil
+}
+
+// Init runs some startup code for this resource.
+func (obj *ConsulKVRes) Init(init *engine.Init) error {
+	obj.init = init // save for later
+
+	obj.config = api.DefaultConfig()
+	if obj.Address != "" {
+		obj.config.Address = obj.Address
+	}
+	if obj.Token != "" {
+		obj.config.Token = obj.Token
+	}
+
+	var err error
+	obj.client, err = api.NewClient(obj.config)
+	return err
+}
+
+// Close is run by the engine to clean up after the resource is done.
+func (obj *ConsulKVRes) Close() error {
+	obj.config.Transport.CloseIdleConnections()
+	return nil
+}
+
+// Watch is the listener and main loop for this resource and it outputs events.
+func (obj *ConsulKVRes) Watch() error {
+	wg := &sync.WaitGroup{}
+	defer wg.Wait()
+
+	c := make(chan error)
+	kv := obj.client.KV()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		opts := &api.QueryOptions{RequireConsistent: true}
+		ctx, cancel := util.ContextWithCloser(context.Background(), obj.init.Done)
```

Comment:
This is a race, because you could call the context close at the same time as the below kv.Get runs.
So you need a separate signal that comes from the below for loop that runs when *it* closes (probably a defer) that then propagates into this goroutine.

Reply:
ok, so I understand what you mean.

But, closing the context while kv.Get runs is what we want. kv.Get will only return when a value has changed (this is the actual watch). So when "done" is called, we want to cancel the running Get.